3scale/3scale-operator
# THREESCALE-10613 add userkey and applicationKey via cr

# What
Issue [THREESCALE-10613](https://issues.redhat.com//browse/THREESCALE-10613)
add the ability to change UserKey and ApplicationKey via CR

# Verification
- install the operator following scripts will do this
[setup_db.sh](https://github.com/3scale-labs/3scale-perf-setup/blob/main/setup_db.sh)
[setup_apimanger.sh](https://github.com/3scale-labs/3scale-perf-setup/blob/main/setup_apimanager.sh)
[setup_product.sh](https://github.com/3scale-labs/3scale-perf-setup/blob/main/setup_product.sh)
- check the application UI to confirm that UserKey and ApplicationKey have been setup (will need to set the product to ApplcitionID & ApplictionKey in the product configuration page)
- scale down the operator deployment
- from this branch do a `make install` and `make run`
### Test generate secret
- Create an secret called auth-secret 
```bash
oc create secret generic auth-secret \
    --from-literal=UserKey=""\
    --from-literal=ApplicationKey=""\
    --namespace=3scale-test
```
- create a applicationAuth CR
```bash
oc apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: ApplicationAuth
metadata:
  name: applicationAuth-cr1
  namespace: 3scale-test
spec:
  applicationCRName: application-cr 
  generateSecret: true
  authSecretRef: 
     name: auth-secret
EOF
```
- confirm in the application ui that the UserKey and ApplicationKey(should see 2 keys) have updated 
- confirm that the auth-secret values have updated to match the latest keys UI and that the ApplicationID is present
### Test manual secret 
- Create an secret called auth-secret2 with the following values
```bash
oc create secret generic auth-secret2 \
    --from-literal=UserKey="testUserKey"\
    --from-literal=ApplicationKey="testApplicationKey"\
    --namespace=3scale-test
```
- create a applicationAuth CR with out genetate secret
```bash
oc apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: ApplicationAuth
metadata:
  name: applicationAuth-cr2
  namespace: 3scale-test
spec:
  applicationCRName: application-cr 
  authSecretRef: 
     name: auth-secret2
EOF
```
- Outcome in the application UI is that the UserKey and ApplictionKey will be populated with the values from the secret 
- The ApplicationID will be added to the auth-secret2
